drivers: can: Implement MCP25xxFD driver

legoabram opened the pull request:
Implementation for the Microchip MCP2517FD/MCP2518FD SPI based CAN-FD controller.

This driver passes the loopback tests presented in the stm32 test, but since there really isn't a standard board with this chip onboard that I'm aware of, I couldn't setup a proper twister test. (Also, the documentation doesn't mention anything about Windows not supporting twister which had me confused for a while.)

Basic rundown, the MCP25XXFD has 32 FIFOs, one of which is a dedicated TX Queue which prioritizes low ID messages. I use this queue for all transmissions, and the size is determined in Kconfig. The TEF is used to know when each transmission ends and to notify the caller using the callback using the SEQ to identify which callback/semaphore to use. The rest of the RAM is dedicated to the largest RX FIFO possible and all filters send messages to that FIFO. Each message is marked with a filter ID so you know which filter it came from.

Since this is my first real contribution, I'm posting this as a draft first to get critique and advice. The one component that I'm unsure of how or haven't gotten to fully test is bus failure or TX failure. I also need to verify the bit timing calculations.

Signed-off-by: Abram Early <abram.early@gmail.com>

legoabram committed:
drivers: can: Implement MCP25xxFD driver

Implementation for Microchip MCP2517FD/MCP2518FD SPI based CAN-FD
controller.

Signed-off-by: Abram Early <abram.early@gmail.com>

karstenkoenig commented:
Hi @legoabram , thanks a lot! I'll have a proper look in the next days, but it will take some time as I'll have to dig into the datasheet for it as well :-)

Are you using an off the shelf mcp25xxfd breakout board like this?
https://wiki.seeedstudio.com/2-Channel-CAN-BUS-FD-Shield-for-Raspberry-Pi/
Then adding that as a shield would also be welcome, as it helps getting started quicker, even if end users might have to clone that shield to adjust to some other pinout

legoabram commented:
> Are you using an off the shelf mcp25xxfd breakout board like this?

Unfortunately, only kind of. I made a custom adapter for an MCP2518FD CLICK which turns it into an arduino shield for dev purposes, but nothing directly off the shelf.

jremmert-phytec-iot commented:
Hi @legoabram, I created a configuration for a shield based on MCP2517FD to test your driver with a reel board. Tested with `samples/subsys/canbus/isotp` and `samples/drivers/can/`, but I can't get any of the examples to work. I don't see activity on the CAN bus yet. Did you tested the driver with any particular setting?

Initially the device initialization failed and I figured that the MODIF interrupt, indicating a successful mode change, comes after about 38 ms. If I extend the timeout for changing MCP25xxFD modes in `mcp25xxfd_set_raw_mode` to 50 ms from 2 ms the initialization seems to work. Some mode changes happen much faster, but a mode change from CONFIGURATION to EXT_LOOPBACK or from CONFIGURATION to NORMAL_CANFD takes longer for me. Did you observed similar behavior?

legoabram commented:
Thanks for the testing @jremmert-phytec-iot. That's really odd... Only thing I can think of is that the documentation states that a mode change requires the bus to be idle before it switches. In which case, I probably should use a much larger timeout like 250ms. I've been testing with this config:
```
can0: mcp2518fd@0 {
	compatible = "microchip,mcp25xxfd";
	spi-max-frequency = <16000000>;
        int-gpios = <&gpio1 5 GPIO_ACTIVE_LOW>;
        status = "okay";
	label = "CAN_0";
	reg = <0>;
	osc-freq = <40000000>;
	sof-on-clko;

        bus-speed = <1000000>;
        bus-speed-data = <8000000>;
		
        sjw = <5>;
	prop-seg = <18>;
	phase-seg1 = <10>;
	phase-seg2 = <11>;
	sjw-data = <1>;
	prop-seg-data = <0>;
	phase-seg1-data = <3>;
	phase-seg2-data = <1>;
       
	#address-cells = <1>;
	#size-cells = <0>;
};
```

I have noticed that if the bit timing is "off", the chip will just give up entirely and lock up transmissions. It will still respond, but the transmissions will never complete. I've had this happen several times when I was testing out different data phase bit timings. I'm still not entirely sure what the conditions are for that failure.

jremmert-phytec-iot commented:
I tested and can confirm that the mode transition time depends on the bus timings. Initially I tested with a bus-speed of 125 k. Still no activity on the can lines, though.

When I set the phase-seg to the values you posted, I get a Kernel Panic. Will try to have a closer look into it in the next days.

talih0 commented on drivers/can/can_mcp25xxfd.h:
```diff
@@ -0,0 +1,530 @@
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	uint32_t RXMSGTS : 32;
+#endif
+	uint8_t DATA[0];
```
Why not just uint8_t DATA[CAN_MAX_DLEN] ?
It would make the code in mcp25xxfd_rx a bit simpler. Same for tx?

legoabram replied:
Okay. Yeah, that's much better. I must have thought I was being clever...

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1084 @@
+#if defined(CONFIG_CAN_FD_MODE)
+	dst->FDF = src->fd;
+#endif
+	memcpy(dst->DATA, src->data, CAN_MAX_DLEN);
```
memcpy(dst->data, src->DATA, src->dlc); ?

Because dst is 
uint8_t objbuf[sizeof(struct mcp25xxfd_txobj) + msg->dlc];

talih0 commented:
Hi @legoabram

Did you have any problems with samples/drivers/can? 
When I tested with the attached changes I get an incorrect payload.
[patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/5922182/patch.txt)
It only seems to work for me if dlc = 4 bytes.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1026 @@
+
+	if (!timing || !timing_data) {
+		return -EINVAL;
+	}
```
For regular CAN frames, *timing_data is parsed as NULL.
You probably need an extra ifdef to check if canfd is enabled then.

legoabram committed:
drivers: can: Timestamps depend on driver select

Instead of explicity ORing together the compatible drivers,
CAN_RX_TIMESTAMP now depends on CAN_HAS_RX_TIMESTAMP
which is selected by the drivers primary Kconfig option. In addition,
stm32fd does not need to select this option since it selected by
the underlying M_CAN driver.

Signed-off-by: Abram Early <abram.early@gmail.com>

legoabram commented:
Hey @talih0, @jremmert-phytec-iot: Would you be willing to take another look at this? I've made some fairly significant updates. I'm still working on an issue where it can lock up if transmissions error out since it isn't quite straightforward to me just yet how to reliably detect and address them.

talih0 reviewed (commented):
Hi,

Just wondering why you changed how interrupt is handled now? (i.e. disabling in the callback and then enabling back).

talih0 commented:
@legoabram 
I've been doing some work to increase throughput. Here's my branch:
https://github.com/talih0/zephyr/commits/mcp25xxfd-driver-github

Feel free to pull in anything you want.

legoabram commented:
> Just wondering why you changed how interrupt is handled now? (i.e. disabling in the callback and then enabling back).

The primary reason is that I was having issues where it would miss an interrupt edge, which would permanently lock up the driver. This seemed to me to be the most reliable solution since it guarantees that the thread will run if the interrupt is asserted. I have to disable the interrupt after it fires since it would just loop endlessly in that callback since the interrupt can't be directly accessed and cleared from an IRQ thread.

> I've been doing some work to increase throughput.

That's exciting! I'll have to take a look. You've got a creative way of addressing the command+address bytes without copying data or using a multi part SPI transaction which introduces a LOT of delay. (Still not totally sure why it takes so long.)

talih0 commented:
> > Just wondering why you changed how interrupt is handled now? (i.e. disabling in the callback and then enabling back).
> 
> The primary reason is that I was having issues where it would miss an interrupt edge, which would permanently lock up the driver. This seemed to me to be the most reliable solution since it guarantees that the thread will run if the interrupt is asserted. I have to disable the interrupt after it fires since it would just loop endlessly in that callback since the interrupt can't be directly accessed and cleared from an IRQ thread.
> 

That does make sense, but I don't know if it's the right way to do it. Hopefully other maintainers can give some advice.

talih0 commented:
> > I've been doing some work to increase throughput.
> 
> That's exciting! I'll have to take a look. You've got a creative way of addressing the command+address bytes without copying data or using a multi part SPI transaction which introduces a LOT of delay. (Still not totally sure why it takes so long.)

I came across the same problem. On my board using DMA in spi driver doubled the throughput. (I'm using XMC4700, but I haven't made a pull request on zephyr yet).
There are also unncessary packet copies in things like CAN message queues. It would nice to improve the api to fix this.
